Custom-Audio-Driver/Custom-Audio-Driver/OTDefaultAudioDevice-Mac.m

```diff
@@ -711,4 +736,202 @@ - (BOOL)setPlayOutRenderCallback:(AudioUnit)unit
+    NSLog(@" - Model UID:       %@", getStringProperty(deviceID, kAudioDevicePropertyModelUID));
+    NSLog(@" - Name:            %@", getStringProperty(deviceID, kAudioDevicePropertyDeviceNameCFString));
+    NSLog(@" - Manufacturer:    %@", getStringProperty(deviceID, kAudioDevicePropertyDeviceManufacturerCFString));
+    NSLog(@" - Input channels:  %@", @(getChannelCount(deviceID, kAudioObjectPropertyScopeInput)));
+    NSLog(@" - Output channels: %@", @(getChannelCount(deviceID, kAudioObjectPropertyScopeOutput)));
+    NSLog(@" - Input source:    %@", getSourceName(deviceID, kAudioObjectPropertyScopeInput));
+    NSLog(@" - Output source:   %@", getSourceName(deviceID, kAudioObjectPropertyScopeOutput));
+    NSLog(@" - Transport type:  %@", getCodeProperty(deviceID, kAudioDevicePropertyTransportType));
+    NSLog(@" - Icon:            %@", getURLProperty(deviceID, kAudioDevicePropertyIcon));
+    return deviceID;
+}
+
+static inline AudioObjectPropertyAddress makeGlobalPropertyAddress(AudioObjectPropertySelector selector)
+{
+    AudioObjectPropertyAddress address = {
+        selector,
+        kAudioObjectPropertyScopeGlobal,
+        kAudioObjectPropertyElementMain,
+
+    };
+    return address;
+}
+
+static NSString *formatStatusError(OSStatus status)
+{
+    if (status == noErr) {
+        return [NSString stringWithFormat:@"No error (%d)", status];
+    }
+
+    return [NSString stringWithFormat:@"Error \"%s\" (%d)",
+            codeToString(status),
+            status];
+}
+
+static char *codeToString(UInt32 code)
+{
+    static char str[5] = { '\0' };
+    UInt32 swapped = CFSwapInt32HostToBig(code);
+    memcpy(str, &swapped, sizeof(swapped));
+    return str;
+}
+
+static NSString *getStringProperty(AudioDeviceID deviceID,
+                                   AudioObjectPropertySelector selector)
+{
+    AudioObjectPropertyAddress address = makeGlobalPropertyAddress(selector);
+    CFStringRef prop;
+    UInt32 propSize = sizeof(prop);
+    OSStatus status = AudioObjectGetPropertyData(deviceID,
+                                                 &address,
+                                                 0,
+                                                 NULL,
+                                                 &propSize,
+                                                 &prop);
+    if (status != noErr) {
+        return formatStatusError(status);
+    }
+    return (__bridge_transfer NSString *)prop;
+}
+
+static NSUInteger getChannelCount(AudioDeviceID deviceID,
+                                  AudioObjectPropertyScope scope)
+{
+    AudioObjectPropertyAddress address = {
+        kAudioDevicePropertyStreamConfiguration,
+        scope,
+        kAudioObjectPropertyElementMain,
+    };
+
+    AudioBufferList streamConfiguration;
+    UInt32 propSize = sizeof(streamConfiguration);
+    OSStatus status = AudioObjectGetPropertyData(deviceID,
```

Comment:
Should we call `AudioObjectGetPropertyDataSize` to know the proper size before allocating streamConfiguration? Or this is enough?

Reply:
On my testing this was enough.